File: apps/sdkwork-im-flutter-mobile/packages/sdkwork_im_flutter_mobile_chat/lib/src/services/chat_message_history_utils.dart
import 'package:sdkwork_im_flutter_mobile_core/sdkwork_im_flutter_mobile_core.dart';

import 'chat_sdk_response_utils.dart';

/// Maximum message history entries retained in memory per conversation (aligned with H5).
const int maxMessageHistoryEntries = 500;

class ChatMessageHistoryResult {
  const ChatMessageHistoryResult({
    required this.items,
    required this.pagination,
  });

  final List<ConversationMessageEntry> items;
  final MessageHistoryPaginationState pagination;
}

class MessageHistoryPaginationState {
  const MessageHistoryPaginationState({
    required this.hasMore,
    required this.nextCursor,
  });

  final bool hasMore;
  final String? nextCursor;
}

enum MessageHistoryWindowDirection { older, newer }

class MessageHistoryPageMergeResult {
  const MessageHistoryPageMergeResult({
    required this.items,
    required this.incomingPageRetained,
  });

  final List<ConversationMessageEntry> items;
  final bool incomingPageRetained;
}

int resolveLatestMessageSeq(List<ConversationMessageEntry> entries) {
  var maxSeq = 0;
  for (final entry in entries) {
    if (entry.messageSeq > maxSeq) {
      maxSeq = entry.messageSeq;
    }
  }
  return maxSeq;
}

MessageHistoryPageMergeResult mergeConversationMessagePage(
  Iterable<ConversationMessageEntry> existing,
  Iterable<ConversationMessageEntry> incoming, {
  required MessageHistoryWindowDirection direction,
}) {
  final existingEntries = existing.toList(growable: false);
  final incomingEntries = incoming.toList(growable: false);
  final existingIds = existingEntries.map((entry) => entry.messageId).toSet();
  final byId = <String, ConversationMessageEntry>{};
  for (final entry in existingEntries) {
    byId[entry.messageId] = entry;
  }
  for (final entry in incomingEntries) {
    byId[entry.messageId] = entry;
  }
  final merged = byId.values.toList()
    ..sort((left, right) {
      final sequenceComparison = left.messageSeq.compareTo(right.messageSeq);
      if (sequenceComparison != 0) {
        return sequenceComparison;
      }
      final occurredAtComparison = left.occurredAt.compareTo(right.occurredAt);
      if (occurredAtComparison != 0) {
        return occurredAtComparison;
      }
      return left.messageId.compareTo(right.messageId);
    });
  final items = direction == MessageHistoryWindowDirection.older
      ? merged.take(maxMessageHistoryEntries).toList(growable: false)
      : merged.length <= maxMessageHistoryEntries
          ? merged
          : merged.sublist(merged.length - maxMessageHistoryEntries);
  final retainedIds = items.map((entry) => entry.messageId).toSet();
  final incomingPageRetained = incomingEntries.every(
    (entry) =>
        existingIds.contains(entry.messageId) ||
        retainedIds.contains(entry.messageId),
  );
  return MessageHistoryPageMergeResult(
    items: items,
    incomingPageRetained: incomingPageRetained,
  );
}

List<ConversationMessageEntry> mergeConversationMessageEntries(
  Iterable<ConversationMessageEntry> existing,
  Iterable<ConversationMessageEntry> incoming, {
  MessageHistoryWindowDirection direction = MessageHistoryWindowDirection.newer,
}) {
  return mergeConversationMessagePage(
    existing,
    incoming,
    direction: direction,
  ).items;
}

MessageHistoryPaginationState readCursorPageInfo(PageInfo? pageInfo) {
  final nextCursor = pageInfo?.nextCursor;
  final hasMore =
      pageInfo?.hasMore == true && nextCursor != null && nextCursor.isNotEmpty;
  return MessageHistoryPaginationState(
    hasMore: hasMore,
    nextCursor: hasMore ? nextCursor : null,
  );
}

MessageHistoryPaginationState pickMessageHistoryPagination(
  ChatMessageHistoryResult? response,
) {
  return response?.pagination ??
      const MessageHistoryPaginationState(hasMore: false, nextCursor: null);
}

ChatMessageHistoryResult readMessageHistoryPageFromSdkResponse(
  ConversationMessageListResponse? response,
) {
  final pageInfo = readPageInfoFromSdkData(response?.data);
  return ChatMessageHistoryResult(
    items: readItemsFromSdkData(
      response?.data,
      ConversationMessageEntry.fromJson,
    ),
    pagination: readCursorPageInfo(pageInfo),
  );
}

PostMessageResult? readPostMessageResultFromSdkResponse(
  ConversationsMessagesCreateResponse201? response,
) {
  return readItemFromSdkData(response?.data, PostMessageResult.fromJson);
}

File: apps/sdkwork-im-h5/packages/sdkwork-im-h5-chat/src/services/chatMessageHistoryUtils.ts
import type { ConversationMessageEntry, SdkWorkListPageInfo } from "@sdkwork/im-sdk";

export const MAX_MESSAGE_HISTORY_ENTRIES = 500;

export interface MessageHistoryPaginationState {
  hasMore: boolean;
  nextCursor?: string;
}

export type MessageHistoryWindowDirection = "older" | "newer";

export interface MessageHistoryPageMergeResult {
  entries: ConversationMessageEntry[];
  incomingPageRetained: boolean;
}

export function readCursorPageInfo(
  pageInfo?: SdkWorkListPageInfo,
): { hasMore: boolean; nextCursor?: string } {
  const nextCursor = pageInfo?.nextCursor;
  const hasMore = pageInfo?.hasMore === true
    && typeof nextCursor === "string"
    && nextCursor.length > 0;
  return {
    hasMore,
    nextCursor: hasMore ? nextCursor : undefined,
  };
}

export function resolveLatestMessageSeq(entries: readonly ConversationMessageEntry[]): number {
  return entries.reduce((max, entry) => Math.max(max, entry.messageSeq ?? 0), 0);
}

export function mergeConversationMessagePage(
  existing: readonly ConversationMessageEntry[],
  incoming: readonly ConversationMessageEntry[],
  direction: MessageHistoryWindowDirection,
): MessageHistoryPageMergeResult {
  const existingIds = new Set(existing.map((entry) => entry.messageId));
  const byId = new Map<string, ConversationMessageEntry>();
  for (const entry of existing) {
    byId.set(entry.messageId, entry);
  }
  for (const entry of incoming) {
    byId.set(entry.messageId, entry);
  }
  const sorted = Array.from(byId.values())
    .sort((left, right) => {
      const sequenceDifference = (left.messageSeq ?? 0) - (right.messageSeq ?? 0);
      if (sequenceDifference !== 0) {
        return sequenceDifference;
      }
      const occurredAtDifference = Date.parse(left.occurredAt) - Date.parse(right.occurredAt);
      if (Number.isFinite(occurredAtDifference) && occurredAtDifference !== 0) {
        return occurredAtDifference;
      }
      return left.messageId.localeCompare(right.messageId);
    });
  const entries = direction === "older"
    ? sorted.slice(0, MAX_MESSAGE_HISTORY_ENTRIES)
    : sorted.slice(-MAX_MESSAGE_HISTORY_ENTRIES);
  const retainedIds = new Set(entries.map((entry) => entry.messageId));
  const incomingPageRetained = incoming.every(
    (entry) => existingIds.has(entry.messageId) || retainedIds.has(entry.messageId),
  );

  return { entries, incomingPageRetained };
}

export function mergeConversationMessageEntries(
  existing: readonly ConversationMessageEntry[],
  incoming: readonly ConversationMessageEntry[],
  direction: MessageHistoryWindowDirection = "newer",
): ConversationMessageEntry[] {
  return mergeConversationMessagePage(existing, incoming, direction).entries;
}

export function pickMessageHistoryPagination(response: {
  pageInfo?: SdkWorkListPageInfo;
}): MessageHistoryPaginationState {
  return readCursorPageInfo(response.pageInfo);
}
